src/lib/types.ts

```diff
@@ -6,50 +6,53 @@ export interface Task<T> {
   cancel: () => void;
 }
 
-export type WaitForAction = <Payload>(actionCreator: ActionCreator<Payload>) => Promise<Action<Payload>>;
+export type WaitForAction = <Payload>(
+  actionCreator: ActionCreator<Payload>,
+) => Promise<Action<Payload>>;
 
-export type FuncWithEnv<State, Args extends any[], T> = (env: SagaEnvironment<State>, ...args: Args) => T;
-
-export abstract class BoundEffect<Env, Params extends any[], ReturnType> {
-  public readonly args: Params;
-
-  constructor(...args: Params) {
-    this.args = args;
-  }
-
-  public abstract run(run: Env, ...args: Params): ReturnType;
-}
+export type BoundFunc<State, Args extends any[], T> = (
+  env: SagaEnvironment<State>,
+  ...args: Args
+) => T;
 
 export interface SagaEnvironment<State> {
   /**
    * Dispatch an action to the store from a saga.
    *
    * @param action - The action object.
    */
-  dispatch(action: Action<any>): void;
+  dispatch<Payload>(action: Action<Payload>): Action<Payload>;
```

Comment:
Is this really true from `redux`? And in what use case would anyone need the action they just put in?

If there is a case for it, I'd suggest to return the `action` rather than `store.dispatch()`'s result.

Reply:
`store.dispatch` actually just returns the action. I thought about potential use-cases for this but couldn't come up with any. I will revert this change again.